### ISSUE TITLE
HWA with Nvidia GPUs + docker

### ISSUE THREAD

**Issue by D34DC3N73R** [jellyfin/jellyfin-docs#273]
I see a couple of issues with the documentation in regards to docker + hwa + nvidia gpus.

1.  Nvidia GPUs are supported when using docker-compose.
Prerequisites: nvidia drivers, nvidia docker 2 package, edited `/etc/docker/daemon.json` to make nvidia the default runtime in docker-compose v3+ (as you have outlined). Simply add the nvidia environment variables to the docker-compose jellyfin service.
```
    environment:
      - NVIDIA_VISIBLE_DEVICES=all
      - NVIDIA_DRIVER_CAPABILITIES=all
      ...
```
This should be compatible with any glibc distribution (which is required for nvidia drivers and nvidia docker 2).

2. When using docker v19.03+ (docker run), there is no reason to edit `/etc/docker/daemon.json` as gpu options have been integrated into the default runtime. The nvidia docker 2 package is no longer needed, but [nvidia container toolkit](https://github.com/NVIDIA/nvidia-docker) is required as well as nvidia drivers. You no longer need the specific nvidia environment variables or switching to the nvidia runtime. Just use  `--gpus all`
```
docker run -d \
--name=jellyfin \
--gpus all \
-p 8096:8096 \
-p 8920:8920 \
-v /config:/config \
-v /media:/media \
-v /cache:/cache \
--restart unless-stopped \
jellyfin/jellyfin
```
Again, this should be compatible with any glibc distribution (which is required for nvidia drivers and nvidia container toolkit).

**Comment by Artiume** [comment]
Would you be willing to get a working config using `user: 1000:1000` and `group_add: videogroup` I know of group 44 and 107 and had issues getting both groups added to the container. It's been hard getting the permissions right

**Comment by D34DC3N73R** [comment]
When using nvidia container toolkit / nvidia docker 2, you do not need to specify or add an alternate video group. This is from my docker-compose.yml. PUID/PGID .env is set to admin:admin (1000). 
```
  jellyfin:
    image: linuxserver/jellyfin
    container_name: jellyfin
    environment:
      - PUID=$PUID
      - PGID=$PGID
      - TZ=$TZ
      - NVIDIA_VISIBLE_DEVICES=all
      - NVIDIA_DRIVER_CAPABILITIES=all
    volumes:
      - $HOME/.config/jellyfin:/config
      - /mnt/media:/media
      - /dev/shm:/transcode
    ports:
      - 8095:8096
    restart: unless-stopped
```
Basically, the only time you need to specify or add the video group to your user is when you're passing /dev/dri (like you would if you wanted to use intel quick sync).

**Comment by Artiume** [comment]
```
environment:
 - PUID=$PUID
 - PGID=$PGID 
```
While I love the lsio image, these aren't compatible with our image lol. So you haven't had issues with setting the user with our image?

**Comment by D34DC3N73R** [comment]
nvenc works with your image as well
```
  jellyfin-test:
    image: jellyfin/jellyfin
    container_name: jellyfin-test
    user: 1000:1000
    environment:
      - TZ=$TZ
      - NVIDIA_VISIBLE_DEVICES=all
      - NVIDIA_DRIVER_CAPABILITIES=all
    volumes:
      - $HOME/.config/jellyfin-test:/config
      - /mnt/media:/media
      - /dev/shm:/transcode
    ports:
      - 8094:8096
```

Here's a snippet from the transcode log.
```
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (h264_cuvid) -> h264 (h264_nvenc))
  Stream #0:1 -> #0:1 (ac3 (native) -> mp3 (libmp3lame))
Press [q] to stop, [?] for help
[hls @ 0x563227fce900] Opening '/transcode/6f1672a3fae3e09591d608543fa5a32c0.ts' for writing
Output #0, hls, to '/transcode/6f1672a3fae3e09591d608543fa5a32c.m3u8':
  Metadata:
    encoder         : Lavf58.29.100
    Stream #0:0: Video: h264 (h264_nvenc) (High), yuv420p, 1280x538 [SAR 1:1 DAR 640:269], q=-1--1, 3360 kb/s, 23.98 fps, 90k tbn, 23.98 tbc (default)
    Metadata:
      encoder         : Lavc58.54.100 h264_nvenc
    Side data:
      cpb: bitrate max/min/avg: 3360000/0/3360000 buffer size: 6720000 vbv_delay: -1
    Stream #0:1: Audio: mp3 (libmp3lame), 48000 Hz, stereo, fltp, 384 kb/s (default)
    Metadata:
      encoder         : Lavc58.54.100 libmp3lame
frame=   77 fps=0.0 q=9.0 size=N/A time=00:00:03.72 bitrate=N/A speed=7.44x    
frame=  217 fps=217 q=10.0 size=N/A time=00:00:09.36 bitrate=N/A speed=9.34x    
[hls @ 0x563227fce900] Opening '/transcode/6f1672a3fae3e09591d608543fa5a32c1.ts' for writing
frame=  355 fps=236 q=10.0 size=N/A time=00:00:15.04 bitrate=N/A speed=  10x    
frame=  485 fps=242 q=25.0 size=N/A time=00:00:20.59 bitrate=N/A speed=10.3x    
```
``

**Comment by Artiume** [comment]
Awesome, I appreciate it.

**Comment by D34DC3N73R** [comment]
I did notice it was using cuvid instead of nvdec, even though hardware decoding is enabled for all supported file types. I figured that was something to do with the jellyfin transcoder rather than docker enabled hardware acceleration. Is nvdec not working in native setups either?

**Comment by Artiume** [comment]
@joshuaboniface you want to take a look into the cuvid and nvdec? I don't know much about nvidia hwa.

**Comment by D34DC3N73R** [comment]
It could be a lack of my understanding. That seems to be the case in emby too. Although it clearly states it's using nvdec.
emby transcoder log snippets
```
>>>>>>  User policy for admin
        EnablePlaybackRemuxing: True
        EnableVideoPlaybackTranscoding: True
        EnableAudioPlaybackTranscoding: True

>>>>>>  Affected codecs
Encoder NVENC Quadro P2000 - H.264 (AVC)
        Adapter #0: 'Quadro P2000' ComputeCapability: 6.1
        Frame Sizes: max 4096x4096 - Width Alignment: 2 - Height Alignment: 2
        Profiles: Baseline Profile (Level 5.2), Main Profile (Level 5.2), High Profile (Level 5.2), High 4:4:4 Predictive Profile (Level 5.2)

Decoder NVDEC Quadro P2000 - H.264 (AVC)
        Adapter #0: 'Quadro P2000' ComputeCapability: 6.1
        Frame Sizes: 48x16...4096x4096 - Width Alignment: 2 - Height Alignment: 2
        Profiles: Constrained Baseline Profile (Level 6), Baseline Profile (Level 6), Main Profile (Level 6), High Profile (Level 6), Constrained High Profile (Level 6), Progressive High Profile (Level 6)


>>>>>>  FindVideoEncoder - Media: h264, UseHardwareCodecs: True, Mode: Automatic
Info    Checking: 'NVENC Quadro P2000 - H.264 (AVC)'
Info    Check successful - selecting 'NVENC Quadro P2000 - H.264 (AVC)'

>>>>>>  FindVideoDecoder - MediaType: h264, Mode: Automatic
Info    Checking: 'NVDEC Quadro P2000 - H.264 (AVC)'
Info    Check successful - selecting 'NVDEC Quadro P2000 - H.264 (AVC)'

Warning: No filter behavior available for filter setsar

>>>>>>  Processing Plan
Info    Name                                        CanDoInHardware      WillDoInHardware     Reason                                  
Info    NVDEC Quadro P2000 - H.264 (AVC)         >> True                 True                 Hardware Codec                           
Info    VideoInput                               >> True                 True                 Matching hardware context                
Info    Scaling                                  >> True                 True                                                          
Info    VideoOutput                              >> True                 True                 Hardware encoder                         
Info    NVENC Quadro P2000 - H.264 (AVC)         >> True                 True                 Hardware Codec                           

>>>>>>  Projected Processing Formats
Info    Previous                HW-Context   Format       SW-Format       Next
Info    h264_cuvid           >> CUDA         cuda         nv12         >> scale_cuda
Info    scale_cuda           >> CUDA         cuda         nv12         >> setsar
Info    setsar               >> CUDA         cuda         nv12         >> 
```
...
```
20:58:26.659 Stream mapping:
20:58:26.659   Stream #0:0 (h264_cuvid) -> scale_cuda (graph 0)
20:58:26.659   setsar (graph 0) -> Stream #0:0 (h264_nvenc)
20:58:26.659   Stream #0:1 -> #0:1 (ac3 (native) -> mp3 (libmp3lame))
20:58:26.659   Stream #0:2 -> #1:0 (subrip (srt) -> webvtt (native))
20:58:26.659   Stream #0:0 -> #1:1 (copy)
20:58:26.659   Stream #0:3 -> #2:0 (subrip (srt) -> webvtt (native))
20:58:26.659   Stream #0:0 -> #2:1 (copy)
20:58:26.659   Stream #0:4 -> #3:0 (subrip (srt) -> webvtt (native))
20:58:26.659   Stream #0:0 -> #3:1 (copy)
20:58:26.659 Press [q] to stop, [?] for help
20:58:27.092 Output #0, segment, to '/config/transcoding-temp/BB536C_%d.ts':
20:58:27.092   Metadata:
20:58:27.092     encoder         : Lavf58.35.100
20:58:27.092     Stream #0:0: Video: h264 (h264_nvenc) (High), cuda, 1280x692 [SAR 1:1 DAR 320:173], q=-1--1, 3616 kb/s, 23.98 fps, 90k tbn, 23.98 tbc (default)
20:58:27.092     Metadata:
20:58:27.092       encoder         : Lavc58.62.100 h264_nvenc
20:58:27.092     Side data:
20:58:27.092       cpb: bitrate max/min/avg: 3616000/0/3616000 buffer size: 7232000 vbv_delay: N/A
20:58:27.092     Stream #0:1: Audio: mp3 (libmp3lame), 48000 Hz, stereo, fltp, 192 kb/s (default)
20:58:27.092     Metadata:
20:58:27.092       encoder         : Lavc58.62.100 libmp3lame
20:58:27.161 frame=    0 fps=0.0 q=0.0 size=       0kB time=00:00:00.00 bitrate=N/A throttle=off speed=   0x    
20:58:27.482 [segment @ 0xd07040] Opening '/config/transcoding-temp/BB536C.m3u8.tmp' for writing
20:58:27.482 SegmentComplete=video:0 Index=2 Start=0.000000 End=9.009022 Duration=9.009022 offset_pts=0 start_pts=0 Frames=72 filename=BB536C_2.ts
20:58:27.482 [segment @ 0xd07040] Opening '/config/transcoding-temp/BB536C_3.ts.tmp' for writing
20:58:27.627 [segment @ 0xd07040] Opening '/config/transcoding-temp/BB536C.m3u8.tmp' for writing
20:58:27.627 SegmentComplete=video:0 Index=3 Start=9.009011 End=12.012022 Duration=3.003011 offset_pts=0 start_pts=9009011 Frames=72 filename=BB536C_3.ts
20:58:27.627 [segment @ 0xd07040] Opening '/config/transcoding-temp/BB536C_4.ts.tmp' for writing
20:58:27.663 frame=  164 fps=163 q=23.0 size=    3273kB time=00:00:13.20 bitrate=2031.4kbits/s throttle=off speed=13.2x    
20:58:27.762 [segment @ 0xd07040] Opening '/config/transcoding-temp/BB536C.m3u8.tmp' for writing
```

**Comment by Artiume** [comment]
https://github.com/jellyfin/jellyfin/issues/482#issuecomment-496012051
